### ISSUE TITLE
Add support for Path.absolute()

### ISSUE THREAD

**Issue by b78688f9-6435-4192-9fe9-2f86f7e61fcf** [python/cpython#73874]
BPO | [29688](https://bugs.python.org/issue29688)
--- | :---
Nosy | @brettcannon, @pitrou, @methane, @serhiy-storchaka, @eryksun, @DimitrisJim, @barneygale
PRs | <li>python/cpython#384</li><li>python/cpython#26153</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2017-03-01.17:37:28.703>
labels = ['3.8', '3.9', '3.10', 'type-feature', 'library', 'docs']
title = 'Add support for Path.absolute()'
updated_at = <Date 2022-01-31.23:20:11.549>
user = 'https://github.com/DimitrisJim'
```

bugs.python.org fields:
```python
activity = <Date 2022-01-31.23:20:11.549>
actor = 'eryksun'
assignee = 'docs@python'
closed = False
closed_date = None
closer = None
components = ['Documentation', 'Library (Lib)']
creation = <Date 2017-03-01.17:37:28.703>
creator = 'Jim Fasarakis-Hilliard'
dependencies = []
files = []
hgrepos = []
issue_num = 29688
keywords = ['patch']
message_count = 20.0
messages = ['288767', '288923', '288931', '289413', '289436', '289452', '289460', '289507', '289509', '289526', '358642', '393731', '395689', '412044', '412099', '412102', '412219', '412221', '412222', '412223']
nosy_count = 9.0
nosy_names = ['brett.cannon', 'pitrou', 'methane', 'docs@python', 'serhiy.storchaka', 'eryksun', 'Jim Fasarakis-Hilliard', '4-launchpad-kalvdans-no-ip-org', 'barneygale']
pr_nums = ['384', '26153']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue29688'
versions = ['Python 3.8', 'Python 3.9', 'Python 3.10']
```

</p></details>

**Comment by b78688f9-6435-4192-9fe9-2f86f7e61fcf** [comment]
Method absolute of Path objects lacked documentation, proposed PR adds relevant method to docs.

**Comment by brettcannon** [comment]
As brought up on the PR, it turns out Path.absolute() is extremely under-tested.

Perhaps we should deprecate Path.absolute() instead of document it and properly test it (and the testing will be necessary to move forward with the documentation)? Path.resolve() handles absolute paths already while also resolving '.' and '..': https://docs.python.org/3/library/pathlib.html#pathlib.Path.resolve. It also works with non-existent paths so unless there's some performance issue I'm not aware of for resolving '.' and '..', then I say we deprecate Path.absolute().

**Comment by brettcannon** [comment]
I've closed the PR on GitHub until we decide whether we just want to deprecate Path.absolute() in favour of Path.resolve().

**Comment by brettcannon** [comment]
I'm still thinking about this but I'm still leaning towards deprecating pathlib.absolute().

**Comment by eryksun** [comment]
resolve() can't replace absolute(). They serve different purposes. Sometimes one wants an absolute path, but without resolving symbolic links. 

absolute() processes a path as a string, which will continue to be true if it's updated to call nt.\_getfullpathname (GetFullPathName) on Windows. OTOH, resolve() can outright fail on Windows. I can write up a list of examples (I can think of 5 or 6 unhandled error codes), but it's not directly relevant to this issue.

**Comment by brettcannon** [comment]
I know it has it's uses (avoiding stat calls is one of them), but it is still undocumented, untested, and has two comments in it saying it needs work. Because of all that it might as well not exist since it doesn't meet our standards of quality.

If someone wants to fix all those issues then we can properly document it as supported, but if no one is willing to then I don't think we should leave unsupported code lying around that people might discover through dir().

And it doesn't serve a \_different_ purpose compared to resolve(), it serves a \_subset_ of resolve()'s purpose since resolve() calls absolute() unconditionally.

**Comment by eryksun** [comment]
What's the rationale for not calling self.\_flavour.pathmod.abspath() to implement absolute()? For example:

```pycon
    >>> p = pathlib.Path('C:/con')
    >>> p._flavour.pathmod.abspath(p)
    '\\\\.\\con'
    >>> p._from_parts((p._flavour.pathmod.abspath(p),), init=False)
    WindowsPath('//./con/')
```


That's almost right except for an unrelated problem that pathlib shouldn't append a trailing slash for \\\\.\\ local device paths. Doing so creates a different path, which may be invalid. \\\\.\\con is a symbolic link to \\Device\\ConDrv\\Console, and adding a trailing backslash after the "Console" filename is invalid. An example where the resulting path is valid but wrong is the volume device \\\\.\\C:, which is a link to something like \\Device\\HarddiskVolume2. Appending a backslash refers to the root directory of the file system on the volume.

**Comment by brettcannon** [comment]
"What's the rationale for not calling self.\_flavour.pathmod.abspath() to implement absolute()?" Beats me. :) It's just how Antoine wrote it and that's all I know.

**Comment by serhiy-storchaka** [comment]
posixpath.abspath() collapses "\<path\>/\<symlink\>/.." to "\<path\>", this is not correct. Not sure about ntpath.abspath().

**Comment by methane** [comment]
> posixpath.abspath() collapses "\<path\>/\<symlink\>/.." to "\<path\>", this is not correct. Not sure about ntpath.abspath().

I feel this is reasonable limitation, and expected behavior for some cases.
So I think adding note about this in document is enough.

**Comment by barneygale** [comment]
The bug Eryk mentioned is logged as #100809. This bug was fixed in #26153. Resolving this issue!
